### DISCUSSION TITLE
What is the best way to remove some objects imported from Helm?

### DISCUSSION THREAD

**Discussion by felixscheinost** [hall/kubenix#14] [Q&A]
Hi, thanks for picking up this project!

I am importing an existing Helm chart, which doesn't have the ability to turn some feature off which I want to turn off.

Currrently I am doing something like that:

```nix
{ kubenix, ... }: let
  helmObjects = (kubenix.evalModules {
    module = { kubenix, ... }: {
      imports = with kubenix.modules; [ k8s helm ];
      kubernetes.helm.releases.test = {
         # [...]
      };
    };
  }).config.kubernetes.objects;
  filteredHelmObjects = builtins.filter (x: true) helmObjects;
in {
  kubernetes.objects = filteredHelmObjects;
}
```

Is that the preferred approach or am I missing something?

**Comment by adrian-gierakowski** [comment]
You can probably set the specific objects to null, if you know their type and name:

`kubernetes.resources.deployments.my-name = mkForce null`

**Reply by felixscheinost** [reply]
I think I tried that but got an error from the module system that `null` wasn't of type `submodule`. Does that make sense?

**Reply by adrian-gierakowski** [reply]
yes, sorry, just checked it and got the same error :(

**Reply by felixscheinost** [reply]
I think from a NixOS perspective it makes sense. There overriding something in this way would probably be considered an anti-pattern.

**Reply by adrian-gierakowski** [reply]
> would probably be considered an anti-pattern

I don't know. I think it could be useful sometimes to be able to "delete" a submodule instance. This might be a good general question to ask on https://discourse.nixos.org/

Coming back to kubenix: maybe we could add an option to helm module which would allow passing a function for filtering k8s object produced by the chart [at the source](https://github.com/hall/kubenix/blob/5bcab30ed204bfbf8ff6a37e67af154ff8aac16e/modules/helm.nix#L122)? This should be quite straightforward.

**Reply by felixscheinost** [reply]
> maybe we could add an option to helm module which would allow passing a function for filtering k8s object produced by the chart at the source

Yeah, I think that would make sense. I am just unsure whether it would fit the API from a design standpoint. From an astethic view, so to say :-) But practically speaking it would make the code I am using right now a lot less ugly.

**Reply by felixscheinost** [reply]
Should I open a PR?

**Reply by adrian-gierakowski** [reply]
I’m not a maintainer but I’d say go for it. I would definitely find it useful sometimes.

**Reply by adrian-gierakowski** [reply]
btw. with this [small change](https://github.com/rhinofi/kubenix/commit/fd905f5323b63faa3e6b6ee3f79c6b01a679f45b) it is be possible to do the following to delete a resource  

```nix
  kubenix.evalModules { modules = [
    {
      imports = [ kubenix.modules.k8s ];
      config.kubernetes.resources.deployments.test = {};
    }
    ({ lib, ... }: {
      config.kubernetes.resources.deployments.test = lib.mkForce null;
      # This also works
      # config.kubernetes.api.resources.apps.v1.Deployment.test = lib.mkForce null;
    })
  ]
```

not sure if there are any places in kubenix where we'd need to guard against the possible `null` value, but I was able to build a moderately complicated project with my fork containing the change (note that my form is still based on the original [kubenix ](https://github.com/xtruder/kubenix) since I haven't migrated yet)

for downstream users this could be a breaking change though, if anyone is iterating overt resources etc

**Reply by hall** [reply]
Thanks for the discussion @felixscheinost and @adrian-gierakowski!

I think filtering and (eventually if it cannot be done at the same time/implementation) being able to arbitrarily modify resources.would be great. Early on, I was considering support for helmfile almost solely to leverage their jsonpatch support but would much rather it be done in a more nix-like way (rather than wrapping yet another tool).

At first glance filtering seems the more generic/safe approach but am open arguments why nullifying might be better or also necessary (e.g., maybe filtering should only _modify_ resources and setting them to `null` would be the answer to deleting with a filter).

Happy to review a PR here.

**Reply by felixscheinost** [reply]
> btw. with this small change it is be possible to do the following to delete a resource

Oh cool! That is indeed only a small change. Hmm, I think I would still rather go with the filtering approach as like you said the chance for breaking something else is lower.

> e.g., maybe filtering should only modify resources and setting them to null would be the answer to deleting with a filter

That's a cool idea. Instead of having a `filter: (Resource) -> Bool` we could have a `map: (Resource) -> Resource?` so returning null would drop the resource and returning non-null would also give the ability to modify the resource. We should probably `mkMerge` the result of the `map` so that it behaves like e.g. `overrideAttrs`, right?

**Reply by hall** [reply]
> the chance for breaking something else is lower

The `README.md` file _does_ say to expect breaking changes :slightly_smiling_face: I'm OK doing so for good design this early on and will create a `CHANGELOG.md` file anyway and start tagging versions beforehand to help mitigate surprises.

> We should probably `mkMerge` the result of the `map` so that it behaves like e.g. `overrideAttrs`, right?

That sounds reasonable. I haven't sat down to think through any sort of implementation yet. Ideally, we could create a few test cases to help ensure functionality matches expectation but I myself am not terribly familiar with the state of testing here so I certainly wouldn't let it slow you down (since we can always go back and add them later).

**Reply by adrian-gierakowski** [reply]
> That's a cool idea. Instead of having a filter: (Resource) -> Bool we could have a map: (Resource) -> Resource? so returning null would drop the resource and returning non-null would also give the ability to modify the resource.

The overriding via merging is already possible since that's what the module system is designed to do. Also there is `overrides` option in the helm module which value is merged with each individual resource produced by the hem chart:

https://github.com/hall/kubenix/blob/b2feaf6b405e38c632c70d65f134ac02a1ba0600/modules/helm.nix#L117-L120

This allows to easily add overrides of props which are common to all produced resources, like `metadata.labels` or `annotations`. For targeting individual resources you can simply add something like the following to you module definitions:

```nix
kubernetes.api.resources.apps.v1.Deployment.test.some.prop.to.override = mkForce value;
```

The only thing that is currently not possible to do is to delete existing resources. To solve this we have a few option:

1. filtering function at the helm chat level (helm module specific)
2. allow resources to be nullified (this would work for any resources, not just ones created with helm module)
3. filtering function at the k8s module level, similar to 1, but could but more general

Here's an example of how the function could be implemented at the [k8s module level](https://github.com/rhinofi/kubenix/commit/189dd184eeedec4a656c9a2aac05a796641bbb5c). I've implemented it some time ago on my fork to allow things like hashing config map names. Note that this modifies the final object list, which means that if you build kubenix manifests together with other modules, which reference kubenix resources (like `kubernetes.resources.deployments.test.metadata.name`), these other modules would not see the modification made by this function.

**Reply by felixscheinost** [reply]
> The overriding via merging is already possible since that's what the module system is designed to do.

You're right. It wouldn't make sense to add another way to do the same thing, just worse ;) 

What if I wanted to try to modify e.g. all `Secret`s though? Would I be able to do that without encountering infinite recursion? I assume that's the reason you implemented `objectPostProcessingFunc` to not operate on the submodules but instead on the final list of objects?


> To solve this we have a few option:

When reading this list I think I would actually like the more generic options most. There's probably no reason to make this Helm specific.

> allow resources to be nullified (this would work for any resources, not just ones created with helm module)

What I like about this option is that it seems more declarative than setting some filter function somewhere, so more in line with the NixOS module system. I don't like `null` values in general though as you now have to consider null values whenever mapping over the `kubernetes.api.resources`.

Would it make sense to instead of allowing nullability add a "synthetic" attribute to all resources, like `kubenix.enabled` or something like that? Which is `true` by default and can be set to `false `to omit the resource?
